feat(BulletList): Add BulletList and Bullet components

jahredhope opened the pull request:
# Change
A Bullet point list is a common text layout pattern used across many sites.
This change provides access to a basic implementation of a bullet point list.

### Indent
The default implementation includes a 15px indent.

### Bottom margin
The default implementation provides a small amount of bottom margin that is common amongst the majority of bullet point lists implementations. An optional property to remove this bottom margin may be required in the future.

### SlideToggle
A missing required property in <SlideToggle /> was causing console errors during development and has been added.

# Example usage

```js
    <BulletPointList>
      <Bullet>Bullet point 1</Bullet>
      <Bullet>Bullet point 2</Bullet>
      <Bullet>Bullet point 3</Bullet>
    </BulletPointList>
```
<img width="150" alt="screen shot 2018-03-16 at 5 16 07 pm" src="https://user-images.githubusercontent.com/13903378/37507126-386faa54-2941-11e8-8189-3175868a4f0a.png">

markdalgleish commented:
Just a few thoughts.

- Do you think we could shorten the component names to `BulletList` and `Bullet`?
- I wonder how this would play out with different font sizes? Might be worth experimenting to make sure the API can handle it. I can imagine us needing larger bullet lists in more promotional material.
- I'm not sure whether it's a good idea to abstract away the `Bullet` components or not. Did you have any thoughts on this?

michaeltaranto commented:
I had similar thoughts around a `List` component with some sort of `type` interface. Allows us to centralise any type of list into here.

jahredhope commented:
Shorten naming:
I preferred the verbose name as I think it's more findable.

Abstract away Bullet:
Happy to pull it in if it's not going to be useful.

Font-sizes:
Currently not supported in this implementation as I wasn't able to find an example where we have non-body font sizes with bullet points. Happy to add support now, but would prefer to wait for an example need.

markdalgleish commented:
Regarding font sizes, I'm more interested in it as a thought experiment to help guide us towards a better API—one that plays better with the other typographical components. I'm concerned the current API is too inflexible, being limited to standard type.

michaeltaranto commented on react/index.js:
```diff
@@ -22,6 +22,8 @@ export { default as Secondary } from './Secondary/Secondary';
 export { default as Strong } from './Strong/Strong';
 export { default as Text } from './Text/Text';
 export { default as TextLink } from './TextLink/TextLink';
+export { default as BulletPoint } from './BulletPoint/BulletPoint';
```
What scenario should we need this exposed? If it's not required at the moment, i would limit increasing the api footprint to just the list for now.

jahredhope commented:
Made some updates based on feedback:
- Renamed BulletPointList to BulletPoint
- Removed the double applying of baseline transform (due to standardText mixin)
- Bullet was initially removed, but with the changes to API to support the above it's been added back in

Keen to get feedback on this new API

markdalgleish commented on react/BulletList/BulletList.sketch.js:
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { BulletList, Bullet } from 'seek-style-guide/react';
+export const symbols = {
```
Thanks for adding some Sketch symbols! 💎👏👏👏

markdalgleish commented on react/BulletList/BulletList.sketch.js:
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { BulletList, Bullet } from 'seek-style-guide/react';
+export const symbols = {
+  'BulletPoint/Small': (
```
These symbol names need to be updated to match the new component name.

markdalgleish commented on react/BulletList/BulletList.test.js:
```diff
@@ -0,0 +1,18 @@
+  it('renders with standard text', () => {
+    const wrapper = shallow(
+      <BulletList>
+        <Text>Foo</Text>
```
Does this test need to be updated to use `Bullet` components?

markdalgleish commented on react/SlideToggle/SlideToggle.sketch.js:
```diff
@@ -4,6 +4,6 @@ import SlideToggle from './SlideToggle';
-  'Slide Toggle/Unchecked': <SlideToggle label="Slide toggle" onChange={noop} checked={false} />,
-  'Slide Toggle/Checked': <SlideToggle label="Slide toggle" onChange={noop} checked={true} />
+  'Slide Toggle/Unchecked': <SlideToggle id="myToggle" label="Slide toggle" onChange={noop} checked={false} />,
+  'Slide Toggle/Checked': <SlideToggle id="myToggle" label="Slide toggle" onChange={noop} checked={true} />
```
This should be a different ID?

markdalgleish commented on react/Text/Text.less:
```diff
@@ -59,3 +59,9 @@
+  display: list-item;
+  list-style-type: disc;
+  margin-left: 1.5 * @gutter-width;
+}
```
So clean! 😍

markdalgleish commented:
Just a few relatively minor comments—this is looking really slick 👍

jahredhope commented:
Thanks @markdalgleish 
Updated to the above comments.

michaeltaranto commented on react/Text/Text.js:
```diff
@@ -13,6 +13,8 @@ import stylesRegular from '../Regular/Regular.less';
 import withTextProps, { SizePropTypes } from '../private/withTextProps';
 
 const Text = ({
+  component,
```
Can we add `component` to the proptypes definition below (along with the default)

jahredhope replied:
Added component proptype definition. Thanks

michaeltaranto commented on react/Text/Text.js:
```diff
@@ -25,35 +27,40 @@ const Text = ({
-    })}>
-    <span
+}) => {
+  const Component = component || (bullet ? 'li' : 'span');
```
If we add `component` to the props list with a default of `span` this can be simplified to: `bullet ? 'li' : component`

jahredhope replied:
This would change the behaviour.
The current behaviour priorises a consumer defined component over a bullet's defaut component.
This was chosen to keep things be a little less magical for the consumer, sometimes honouring a custom component and sometimes not may lead to confusion while debugging.

michaeltaranto commented on react/Text/Text.less:
```diff
@@ -59,3 +59,9 @@
+.bullet {
+  display: list-item;
+  list-style-type: disc;
+  margin-left: 1.5 * @gutter-width;
```
I dont like what im about to suggest, but, if we pull this out to a mixin we can use it in places where we cannot use the style guide components themselves, not dissimilar to the text mixins. Thoughts?

markdalgleish replied:
YAGNI (IMO) 😄

michaeltaranto replied:
We already do, i'll take it offline.

jahredhope replied:
I'd rather avoid supporting consumers using mixins without an known consumer.
We could pull this out to a mixin if and when it is required.

Thats just a preference, if you want to support mixin consumers from day one I will.

michaeltaranto replied:
There's already a need for it, it's not hypothetical. Check slack 😄
